xain/grpc/coordinator.py

```diff
@@ -113,102 +120,144 @@ def monitor_heartbeats(participants, terminate_event):
+
+    def Heartbeat(self, request, context):
+        return self.coordinator.on_message(request, context.peer())
+
+    def StartTraining(self, request, context):
+        return self.coordinator.on_message(request, context.peer())
+
+    def EndTraining(self, request, context):
+        return self.coordinator.on_message(request, context.peer())
+
+
+class Coordinator:
+    def __init__(self, required_participants=10, theta=None, epochs=0, epoch_base=0):
         self.required_participants = required_participants
-        self.participants = participants
-        # global model data (sent)
+        self.participants = Participants()
+
+        # global model
         self.theta = [] if theta is None else theta
         self.epochs = epochs
         self.epoch_base = epoch_base
-        # local model data (received)
-        self.theta_updates: List[Tuple[Theta, int]] = []
-        self.histories: List[History] = []
-        self.metricss: List[Metrics] = []
 
-    def Rendezvous(self, request, context):
-        if self.participants.len() < self.required_participants:
-            response = coordinator_pb2.RendezvousResponse.ACCEPT
-            self.participants.add(context.peer())
-            print(
-                f"Accepted participant {context.peer()}"
-                f" # participants: {self.participants.len()}"
-            )
-        else:
-            response = coordinator_pb2.RendezvousResponse.LATER
-            print(
-                f"Rejected participant {context.peer()}"
-                f" # participants: {self.participants.len()}"
+        # local model
+        self.theta_updates = []
+        self.histories = []
+        self.metricss = []
+
+        # state variables
+        self.state = STANDBY
+        self.round = 0
+
+    def on_message(self, message, peer_id):
+        print(f"Received: {type(message)} from {peer_id}")
+
+        if type(message) == coordinator_pb2.RendezvousRequest:
+            # Handle rendezvous
+
+            if self.participants.len() < self.required_participants:
+                response = coordinator_pb2.RendezvousResponse.ACCEPT
+                self.participants.add(peer_id)
+                print(
+                    f"Accepted participant {peer_id}"
+                    f" # participants: {self.participants.len()}"
+                )
+            else:
+                # If state is STANDBY start a round
+
+                response = coordinator_pb2.RendezvousResponse.LATER
+                print(
+                    f"Rejected participant {peer_id}"
+                    f" # participants: {self.participants.len()}"
+                )
+
+            # Change the state to ROUND if we are in STANDBY and already have enough participants
+            if (
+                self.state == STANDBY
+                and self.participants.len() == self.required_participants
+            ):
+                # TODO: We may need to make this update thread safe
+                self.state = ROUND
```

Comment:
maybe change state in the `if` block above (where we should be guaranteed to be in STANDBY)? If we already have `required_participants` we shouldn't need to check this.